uftpd.8
.\"
.\" Copyright (c) 2014-2017  Joachim Nilsson <troglobit@gmail.com>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd Sep 03, 2017
.Dt UFTPD 8
.Os "uftpd (2.4)"
.Sh NAME
.Nm uftpd
.Nd
No nonsense TFTP/FTP Server
.Sh SYNOPSIS
.Nm
.Op Fl hnsv
.Op Fl l Ar LVL
.Op Fl o Ar ftp=PORT,tftp=PORT
.Op Ar PATH
.Sh DESCRIPTION
.Nm
is a very simple TFTP and anonymous FTP server with inetd support.  It
listens on standard Internet ports for each protocol, as defined in the
system service specification,
.Pa /etc/services ,
unless other ports are given on the command line.  For details, see
.Xr services 5 .
.Pp
Without any command line arguments
.Nm
serves both FTP and TFTP and automatically backgrounds itself.  Usually
.Nm
this means listen to port 21 (FTP) and port 69 (TFTP), serve files from
.Pa /srv/ftp ,
and log anything out of the ordinary to syslog.  Messages are written to
the syslog using the
.Nm LOG_FTP
facility.
.Pp
Available command line options:
.Bl -tag -width Ds
.It Fl h
Show built-in help text
.It Fl l Ar LVL
Set log level: none, err, info,
.Ar notice ,
debug
.It Fl n
Run in foreground, do not detach from controlling terminal
.It Fl o
Set
.Nm
option:
.Bl -tag
.It Ar ftp=PORT
.It Ar tftp=PORT
.It Ar writable
.El
.Pp
Override Internet ports otherwise derived from
.Xr services 5 .
Set the
.Ar PORT
to zero (0) to disable a service.
.Pp
The
.Ar writable
option is to enable writable FTP root, which is not recommended.  Some
people want this, but it is recommended to instead rely on a writable
sub-directory, like
.Ar upload/ ,
or similar.
.It Fl s
Use syslog, even if running in foreground, default when running in the
background
.It Fl v
Show program version
.It Ar PATH
Root directory. The default is to serve files from the FTP user's $HOME
.El
.Pp
.Sh Inetd
.Nm
can also be used with an Internet superserver, like the traditional
inetd or modern init replacements like finit.  In inetd mode the server
takes client connections from stdin.  To enable inetd mode
.Nm
must be called as either
.Nm in.tftpd
or
.Nm in.ftpd .
In inetd mode
.Nm
always runs in the foreground with syslog for messages.
.Pp
.Sh FTP
The file
.Pa /etc/nologin
can be used to disable FTP access.  If the file exists,
.Nm
displays it and exits.  If the file
.Pa /etc/ftpwelcome
exists,
.Nm
prints it before issuing the
.Dq ready
message.
If the file
.Pa /etc/motd
exists,
.Nm
prints it after a successful login.  If the file
.Pa .message
exists in a directory,
.Nm
prints it when that directory is entered.
.Pp
The FTP server currently supports the following requests.
The case of the requests is ignored.
.Bl -column "Request" -offset indent
.It Request Ta "Description"
.It ABOR Ta "abort current transfer"
.It CDUP Ta "shorthand for CD .. command"
.It CWD Ta "change working directory"
.It DELE Ta "delete a file"
.It EPSV Ta "extended PASV command, used by VLC for Android"
.It LIST Ta "give list files in a directory" Pq Dq Li "ls -lgA"
.It NLST Ta "like LIST, but much less verbose"
.It MLST Ta "RFC 3659 extension to LIST"
.It MLSD Ta "RFC 3659 extension to LIST"
.It MKD Ta "make a directory"
.It NOOP Ta "do nothing, used for keep-alive"
.It PASS Ta "specify password"
.It PASV Ta "prepare for server-to-server transfer"
.It PORT Ta "specify data connection port"
.It PWD Ta "print the current working directory"
.It QUIT Ta "terminate session"
.It REST Ta "restore RETR or STOR command at file offset"
.It RETR Ta "retrieve a file"
.It RMD Ta "remove a directory"
.It RNFR Ta "specify rename-from file name"
.It RNTO Ta "specify rename-to file name"
.It SIZE Ta "return size of file"
.It STOR Ta "store a file"
.It SYST Ta "show operating system type of server system"
.It TYPE Ta "specify data transfer" Em type
.It USER Ta "specify user name"
.El
.Pp
Remaining FTP requests, as specified in Internet RFC959, are not
recognized at the moment.  Patches are welcome!
.Pp
.Sh TFTP
.Nm
also supports TFTP, the Trivial File Transfer Protocol, which is
often used for net booting diskless devices, e.g., BOOTP and PXEBOOT.
.Pp
The TFTP server currently supports the following requests.
.Bl -column "Request" -offset indent
.It RRQ
.It ERROR
.It ACK
.El
.Pp
.Nm
supports TFTP blocksize negotiation, according to RFC2348.
Support for WRQ is not yet implemented, patches welcome!
.Pp
.Sh FILES
.Bl -tag -width /etc/ftpwelcome -compact
.It Pa /etc/ftpwelcome
FTP Welcome notice.
.It Pa /etc/motd
Message of the day, presented after successful FTP login.
.It Pa /etc/nologin
Displayed to user attempting to connect.  Access is refused if this
file exists.
.El
.Sh SEE ALSO
.Xr ftp 1 ,
.Xr tftp 1 ,
.Xr syslogd 8
.Sh AUTHORS
.Nm
was written by Joachim Nilsson
.Aq mailto:troglobit@gmail.com
and is maintained at
.Aq https://github.com/troglobit/uftpd
.Sh BUGS
Here be dragons.


CHANGELOG.md
Change Log
==========

All notable changes to the project are documented in this file.


[v1.9][] - 2015-07-23
---------------------

Bug fix release.  FTP and TFTP sessions can now run fully in parallel,
independent of each other.  Also improved compatibility with Firefox
built-in FTP client and wget.

### Changes
- Upgrade to [libuEv][] v1.2.1+ for improved error handling and a much
  cleaner API.
- Major refactor of both FTP and TFTP servers to use libuEv better.
- Move to use [libite][] v1.0.0 for `strlcpy()`, `strlcat()`, `pidfile()`
  and more.
- Add proper session timout to TFTP, like what FTP already has.
- Add support for `NLST` FTP command, needed for multiple get operations.
  This fixes issue #2, thanks to @oz123 on GitHub for pointing this out!
- Add support for `FEAT` and `HELP` FTP commands used by some clients.

### Fixes
- Fix issue #3: do not sleep 2 sec before exiting.  Simply forward the
  `SIGTERM` to any FTP/TFTP session in progress, yield the CPU to let
  the child sessions handle the signal, and then exit.  Much quicker!
- Fix issue #4: due to an ordering bug between the main process calling
  `daemon()` and `sig_init()`, we never got the `SIGCHILD` to be able to
  reap any exiting FTP/TFTP sessions.  This resulted in zombies(!) when
  *not* being called as `uftpd -n`
- Fix issue #5: `LIST` and `NLST` ignores path argument sent by client.
- Fix issue #6: FTP clients not detecting session timeout.  Caused by
  uftpd not performing a proper `shutdown()` on the client socket(s)
  before `close()`.
- Fix problem with [libuEv][] not being properly cleaned on `distclean`.
- Fix problem with uftpd not exiting client session properly when client
  simply closes the connection.


[v1.8][] - 2015-02-02
---------------------

### Changes
- Updated [README.md][]
- Add [TODO.md][]
- Add [CHANGELOG.md][], attempt to align with <http://keepachangelog.com>
- From now on [Travis-CI][] only runs when pushing to the dev branch,
  so all new development must be done there.
- Upgrade to [libuEv][] v1.0.4

### Fixes
- Fix insecure `chroot()` reported in [Coverity Scan][] CID #54523.
- Minor cleanup fixes.


[v1.7][] - 2014-12-21
---------------------

The TFTP Blocksize Negotiation release.

### Changes
- Support for [RFC 2348][], TFTP blocksize negotiation
- Support for custom server directory, instead of FTP user's `$HOME`
- Log to `stderr` when running in foreground and debug is enabled


[v1.6][] - 2014-09-12
---------------------

Fix missing [libuEv][] directory content generated by <kbd>make dist</kbd>
in [v1.3][], [v1.4][], and [v1.5][].

### Fixes
- Since the introduction of the event library [libuEv][] the <kbd>make
  dist</kbd> target has failed to include the libuev sub-directory.
  This is due to the `git archive` command unfortunately not supporting
  git sub-modules.


[v1.5][] - 2014-09-12 [YANKED]
------------------------------

Major fix release, lots of issues reported by [Coverity Scan][] fixed.
For details, see <https://scan.coverity.com/projects/2947>

**Note:** This release has been *yanked* from distribution due to the
tarball (generated by the <kbd>make dist</kbd>) missing the required
libuEv library.  Instead, use [v1.6][] or later, where this is fixed, or
roll your own build of this release from the GIT source tree.

### Changes
- Add support for [Travis-CI][], continuous integration with GitHub
- Add support for [Coverity Scan][], the best static code analyzer,
  integrated with [Travis-CI][] -- scan runs for each push to master

### Fixes
- Fix nasty invalid `sizeof()` argument to `recv()` causing uftpd to
  only read 4/8 bytes (32/64 bit arch) at a time from the FTP socket.
  This should greatly reduce CPU utilisation and improve xfer speeds.
  Found by [Coverity Scan][].
- Fix minor resource leak in `ftp_session()` when `getsockname()` or
  `getpeername()` fail.  Minor fix because the session exits and the OS
  usually frees resources at that point, unless you're using uClinux.
  Found by [Coverity Scan][].
- Various fixes for unchecked API return values, prevents propagation of
  errors.  Also, make sure to clear input data before calling API's.
  Found by [Coverity Scan][].
- Fix oversight in checking for invalid/missing FTP username.
  Found by [Coverity Scan][].
- Fix potential attack vector.  Make sure to always store a NUL string
  terminator in all received FTP commands so the parser does not go out
  of bounds. Found by [Coverity Scan][].
- Fix parallel build problems in `Makefile`.


[v1.4][] - 2014-09-04 [YANKED]
------------------------------

**Note:** This release has been *yanked* from distribution due to the
tarball (generated by the <kbd>make dist</kbd>) missing the required
libuEv library.  Instead, use [v1.6][] or later, where this is fixed, or
roll your own build of this release from the GIT source tree.

### Changes
- Update documentation, both built-in usage text and man page.

### Fixes
- Fix bug in inetd.conf installed by .deb package for TFTP service.
  Inetd forked off a new TFTP session for each connection attempt.


[v1.3][] - 2014-09-04 [YANKED]
------------------------------

Added support for TFTP, [RFC 1350][].  Integration of the asynchronous
event library [libuEv][], to serialize all events.  Massive refactoring.

**Note:** This release has been *yanked* from distribution due to the
tarball (generated by the <kbd>make dist</kbd>) missing the required
libuEv library.  Instead, use [v1.6][] or later, where this is fixed, or
roll your own build of this release from the GIT source tree.

### Changes
- Incompatible changes to the command line arguments, compared to v1.2!
- Add libuEv as a GIT submodule, handles signals, timers, and all I/O.
- Refactor all signal handling, timers, and socket `poll()` calls to
  use libuEv instead.  Much cleaner and maintaiable code as a result.
- Clarify copyright claims, not much remains of the original [FtpServer][]
  code, by [Xu Wang][].


[v1.2][] - 2014-05-19
---------------------


src/uftpd.c
/* uftpd -- the no nonsense (T)FTP server
 *
 * Copyright (c) 2014-2018  Joachim Nilsson <troglobit@gmail.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "uftpd.h"

/* Global daemon settings */
char *prognm      = PACKAGE_NAME;
char *home        = NULL;
int   inetd       = 0;
int   background  = 1;
int   do_syslog   = 1;
int   do_ftp      = FTP_DEFAULT_PORT;
int   do_tftp     = TFTP_DEFAULT_PORT;
int   do_insecure = 0;
pid_t tftp_pid    = 0;
struct passwd *pw = NULL;

/* Event contexts */
static uev_t ftp_watcher;
static uev_t tftp_watcher;
static uev_t sigchld_watcher;
static uev_t sigterm_watcher;
static uev_t sigint_watcher;
static uev_t sighup_watcher;
static uev_t sigquit_watcher;


static int version(void)
{
	printf("%s\n", PACKAGE_VERSION);
	return 0;
}

static int usage(int code)
{
	int is_inetd = string_match(prognm, "in.");

	if (is_inetd)
		printf("\nUsage: %s [-hv] [-l LEVEL] [PATH]\n\n", prognm);
	else
		printf("\nUsage: %s [-hnsv] [-l LEVEL] [-o ftp=PORT,tftp=PORT] [PATH]\n\n", prognm);

	printf("  -h         Show this help text\n"
	       "  -l LEVEL   Set log level: none, err, info, notice (default), debug\n");
	if (!is_inetd)
		printf("  -n         Run in foreground, do not detach from controlling terminal\n"
		       "  -o OPT     Options:\n"
		       "                      ftp=PORT\n"
		       "                      tftp=PORT\n"
		       "                      writable\n"
		       "  -s         Use syslog, even if running in foreground, default w/o -n\n");

	printf("  -v         Show program version\n\n");
	printf("The optional 'PATH' defaults to the $HOME of the /etc/passwd user 'ftp'\n"
	       "Bug report address: %-40s\n\n", PACKAGE_BUGREPORT);

	return code;
}

/*
 * SIGCHLD: one of our children has died
 */
static void sigchld_cb(uev_t *w, void *arg, int events)
{
	while (1) {
		pid_t pid;

		pid = waitpid(0, NULL, WNOHANG);
		if (pid <= 0)
			break;

		/* TFTP client disconnected, we can now serve TFTP again! */
		if (pid == tftp_pid) {
			DBG("Previous TFTP session ended, restarting TFTP watcher ...");
			tftp_pid = 0;
			uev_io_start(&tftp_watcher);
		}
	}
}

/*
 * SIGQUIT: request termination
 */
static void sigquit_cb(uev_t *w, void *arg, int events)
{
	INFO("Recieved signal %d, exiting ...", w->signo);

	/* Forward signal to any children in this process group. */
	if (killpg(getpgrp(), SIGTERM))
		WARN(errno, "Failed signalling children");

	/* Give them time to exit gracefully. */
	while (wait(NULL) != -1)
		;

	/* Leave main loop. */
	uev_exit(w->ctx);
}

static void sig_init(uev_ctx_t *ctx)
{
	uev_signal_init(ctx, &sigchld_watcher, sigchld_cb, NULL, SIGCHLD);
	uev_signal_init(ctx, &sigterm_watcher, sigquit_cb, NULL, SIGTERM);
	uev_signal_init(ctx, &sigint_watcher,  sigquit_cb, NULL, SIGINT);
	uev_signal_init(ctx, &sighup_watcher,  sigquit_cb, NULL, SIGHUP);
	uev_signal_init(ctx, &sigquit_watcher, sigquit_cb, NULL, SIGQUIT);
}

static int find_port(char *service, char *proto, int fallback)
{
	int port = fallback;
	struct servent *sv;

	sv = getservbyname(service, proto);
	if (!sv)
		WARN(errno, "Cannot find service %s/%s, defaulting to %d.", service, proto, port);
	else
		port = ntohs(sv->s_port);

	DBG("Found port %d for service %s, proto %s (fallback port %d)", port, service, proto, fallback);

	return port;
}

/*
 * Check that we don't have write access to the FTP root,
 * unless explicitly allowed
 */
static int security_check(char *home)
{
	if (access(home, F_OK)) {
		ERR(errno, "Cannot access FTP root %s", home);
		return 1;
	}

	if (!do_insecure && access(home, W_OK)) {
		ERR(0, "FTP root %s writable, possible security violation!", home);
		return 1;
	}

	return 0;
}

static int init(uev_ctx_t *ctx)
{
	/* Figure out FTP/TFTP ports */
	if (do_ftp == 1)
		do_ftp  = find_port(FTP_SERVICE_NAME, FTP_PROTO_NAME, FTP_DEFAULT_PORT);
	if (do_tftp == 1)
		do_tftp = find_port(TFTP_SERVICE_NAME, TFTP_PROTO_NAME, TFTP_DEFAULT_PORT); 

	/* Figure out FTP home directory */
	if (!home) {
		pw = getpwnam(FTP_DEFAULT_USER);
		if (!pw) {
			WARN(errno, "Cannot find user %s, falling back to %s as FTP root.",
			     FTP_DEFAULT_USER, FTP_DEFAULT_HOME);
			home = strdup(FTP_DEFAULT_HOME);
		} else {
			home = strdup(pw->pw_dir);
		}
	}

	if (!home || security_check(home))
		return 1;

	return uev_init(ctx);
}

static void ftp_cb(uev_t *w, void *arg, int events)
{
        int client;

	if (UEV_ERROR == events || UEV_HUP == events) {
		uev_io_stop(w);
		close(w->fd);
		return;
	}

        client = accept(w->fd, NULL, NULL);
        if (client < 0) {
                WARN(errno, "Failed accepting FTP client connection");
                return;
        }

        ftp_session(arg, client);
}

static void tftp_cb(uev_t *w, void *arg, int events)
{
	uev_io_stop(w);

	if (UEV_ERROR == events || UEV_HUP == events) {
		close(w->fd);
		return;
	}

        tftp_pid = tftp_session(arg, w->fd);
	if (tftp_pid < 0) {
		tftp_pid = 0;
		uev_io_start(w);
	}
}

static int start_service(uev_ctx_t *ctx, uev_t *w, uev_cb_t *cb, int port, int type, char *desc)
{
	int sd;

	if (!port)
		/* Disabled */
		return 1;

	sd = open_socket(port, type, desc);
	if (sd < 0) {
		if (EACCES == errno)
			WARN(0, "Not allowed to start %s service.%s",
			     desc, port < 1024 ? "  Privileged port." : "");
		return 1;
	}

	INFO("Starting %s server on port %d ...", desc, port);
	uev_io_init(ctx, w, cb, ctx, sd, UEV_READ);

	return 0;
}

static int serve_files(uev_ctx_t *ctx)
{
	int ftp, tftp;

	DBG("Starting services ...");
	ftp  = start_service(ctx, &ftp_watcher,   ftp_cb, do_ftp, SOCK_STREAM, "FTP");
	tftp = start_service(ctx, &tftp_watcher, tftp_cb, do_tftp, SOCK_DGRAM, "TFTP");

	/* Check if failed to start any service ... */
	if (ftp && tftp)
		return 1;

	/* Setup signal callbacks */
	sig_init(ctx);

	/* We're now up and running, save pid file. */
	pidfile(NULL);

	INFO("Serving files from %s ...", home);

	return uev_run(ctx, 0);
}

static char *progname(char *arg0)
{
       char *nm;

       nm = strrchr(arg0, '/');
       if (nm)
	       nm++;
       else
	       nm = arg0;

       return nm;
}

int main(int argc, char **argv)
{
	int c;
	enum {
		FTP_OPT = 0,
		TFTP_OPT,
		SEC_OPT,
	};
	char *subopts;
	char *const token[] = {
		[FTP_OPT]  = "ftp",
		[TFTP_OPT] = "tftp",
		[SEC_OPT]  = "writable",
		NULL
	};
	uev_ctx_t ctx;

	prognm = progname(argv[0]);
	while ((c = getopt(argc, argv, "hl:no:sv")) != EOF) {
		switch (c) {
		case 'h':
			return usage(0);

		case 'l':
			loglevel = loglvl(optarg);
			if (-1 == loglevel)
				return usage(1);
			break;

		case 'n':
			background = 0;
			do_syslog--;
			break;

		case 'o':
			subopts = optarg;
			while (*subopts != '\0') {
				char *value;

				switch (getsubopt(&subopts, token, &value)) {
				case FTP_OPT:
					if (!value) {
						fprintf(stderr, "Missing port argument to -o ftp=PORT\n");
						return usage(1);
					}
					do_ftp = atoi(value);
					break;

				case TFTP_OPT:
					if (!value) {
						fprintf(stderr, "Missing port argument to -o tftp=PORT\n");
						return usage(1);
					}
					do_tftp = atoi(value);
					break;

				case SEC_OPT:
					do_insecure = 1;
					break;

				default:
					fprintf(stderr, "Unrecognized option '%s'\n", value);
					return usage(1);
				}
			}
			break;

		case 's':
			do_syslog++;
			break;

		case 'v':
			return version();

		default:
			return usage(1);
		}
	}

	if (optind < argc) {
		size_t len;

		home = strdup(argv[optind]);
		if (!home) {
			ERR(errno, "Failed allocating memory");
			return 1;
		}

		len = strlen(home) - 1;
		if (home[len] == '/')
			home[len] = 0;
	}

	/* Inetd mode enforces foreground and syslog */
	if (string_compare(prognm, "in.tftpd")) {
		inetd      = 1;
		do_ftp     = 0;
		do_tftp    = 1;
		background = 0;
		do_syslog  = 1;
	} else if (string_compare(prognm, "in.ftpd")) {
		inetd      = 1;
		do_ftp     = 1;
		do_tftp    = 0;
		background = 0;
		do_syslog  = 1;
	}

	if (do_syslog) {
		openlog(prognm, LOG_PID | LOG_NDELAY, LOG_FTP);
		setlogmask(LOG_UPTO(loglevel));
	}

	DBG("Initializing ...");
	if (init(&ctx)) {
		ERR(0, "Failed initializing, exiting.");
		return 1;
	}

	if (inetd) {
		int sd;
		pid_t pid;

		INFO("Started from inetd, serving files from %s ...", home);

		/* Ensure socket is non-blocking */
		sd = STDIN_FILENO;
		(void)fcntl(sd, F_SETFL, fcntl(sd, F_GETFL, 0) | O_NONBLOCK);

		if (do_tftp)
			pid = tftp_session(&ctx, sd);
		else
			pid = ftp_session(&ctx, sd);

		if (-1 == pid)
			return 1;
		return 0;
	}

	if (background) {
		DBG("Daemonizing ...");
		if (-1 == daemon(0, 0)) {
			ERR(errno, "Failed daemonizing");
			return 1;
		}
	}

	DBG("Serving files as PID %d ...", getpid());
	return serve_files(&ctx);
}

/**
 * Local Variables:
 *  indent-tabs-mode: t
 *  c-file-style: "linux"
 * End:
 */
